File: 78-subsets/subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(nums, 0, new ArrayList<>(), res);
        return res;
    }
    void dfs(int[] nums, int l, List<Integer> curr, List<List<Integer>> res) {
        if (l == nums.length) {
            res.add(new ArrayList<>(curr));
            return;
        }
        curr.add(nums[l]);
        dfs(nums, l + 1, curr, res);
        curr.remove(curr.size() - 1);
        dfs(nums, l + 1, curr, res);
    }
}

File: 78-subsets/subsets.py
class Solution(object):
    def dfs(self, nums, l, curr, res):
        if l == len(nums):
            res.append(curr[:])
            return
        curr.append(nums[l])
        self.dfs(nums, l + 1, curr, res)
        curr.pop()
        self.dfs(nums, l + 1, curr, res)
    def subsets(self, nums):
        res = []
        self.dfs(nums, 0, [], res)
        return res